pdns/pdns_recursor.cc

```diff
@@ -4633,61 +4816,45 @@ void parseACLs()
-	throw runtime_error("Unable to re-parse configuration file include '"+fn+"'");
+        throw runtime_error("Unable to re-parse configuration file include '"+fn+"'");
+
+      if(!::arg().preParseFile(fn.c_str(), "allow-notify-from-file", ::arg()["allow-notify-from-file"]))
+        throw runtime_error("Unable to re-parse configuration file include '"+fn+"'");
+      if(!::arg().preParseFile(fn.c_str(), "allow-notify-from", ::arg()["allow-notify-from"]))
+        throw runtime_error("Unable to re-parse configuration file include '"+fn+"'");
     }
 
     ::arg().preParse(g_argc, g_argv, "allow-from-file");
     ::arg().preParse(g_argc, g_argv, "allow-from");
-  }
 
-  std::shared_ptr<NetmaskGroup> oldAllowFrom = t_allowFrom;
-  std::shared_ptr<NetmaskGroup> allowFrom = std::make_shared<NetmaskGroup>();
-
-  if(!::arg()["allow-from-file"].empty()) {
-    string line;
-    ifstream ifs(::arg()["allow-from-file"].c_str());
-    if(!ifs) {
-      throw runtime_error("Could not open '"+::arg()["allow-from-file"]+"': "+stringerror());
-    }
+    ::arg().preParse(g_argc, g_argv, "allow-notify-from-file");
+    ::arg().preParse(g_argc, g_argv, "allow-notify-from");
+  }
 
-    string::size_type pos;
-    while(getline(ifs,line)) {
-      pos=line.find('#');
-      if(pos!=string::npos)
-        line.resize(pos);
-      boost::trim(line);
-      if(line.empty())
-        continue;
+  auto oldAllowFrom = t_allowFrom;
 
-      allowFrom->addMask(line);
-    }
-    g_log<<Logger::Warning<<"Done parsing " << allowFrom->size() <<" allow-from ranges from file '"<<::arg()["allow-from-file"]<<"' - overriding 'allow-from' setting"<<endl;
-  }
-  else if(!::arg()["allow-from"].empty()) {
-    vector<string> ips;
-    stringtok(ips, ::arg()["allow-from"], ", ");
+  auto allowFrom = parseACL("allow-from-file", "allow-from");
 
-    g_log<<Logger::Warning<<"Only allowing queries from: ";
-    for(vector<string>::const_iterator i = ips.begin(); i!= ips.end(); ++i) {
-      allowFrom->addMask(*i);
-      if(i!=ips.begin())
-        g_log<<Logger::Warning<<", ";
-      g_log<<Logger::Warning<<*i;
-    }
-    g_log<<Logger::Warning<<endl;
-  }
-  else {
+  if(allowFrom->size() == 0) {
     if(::arg()["local-address"]!="127.0.0.1" && ::arg().asNum("local-port")==53)
       g_log<<Logger::Warning<<"WARNING: Allowing queries from all IP addresses - this can be a security risk!"<<endl;
     allowFrom = nullptr;
   }
 
   g_initialAllowFrom = allowFrom;
-  broadcastFunction([=]{ return pleaseSupplantACLs(allowFrom); });
+  broadcastFunction([=]{ return pleaseSupplantAllowFrom(allowFrom); });
   oldAllowFrom = nullptr;
 
+  auto oldAllowNotifyFrom = t_allowNotifyFrom;
+
+  auto allowNotifyFrom = parseACL("allow-notify-from-file", "allow-notify-from");
+
+  g_initialAllowNotifyFrom = allowNotifyFrom;
+  broadcastFunction([=]{ return pleaseSupplantAllowNotifyFrom(allowNotifyFrom); });
+  oldAllowNotifyFrom = nullptr;
```

Comment:
`oldAllowNotifyFrom` seems unused

Reply:
Indeed, and this pattern was duplicated from the block above for `allow-from`. The only assumption I can make is that this ensures that the object's destructor is called in *this* thread and not in any of the worker threads who are replacing their copies of the `shared_ptr`. If that's not the reason, then the reason is unclear, or lost to the mists of time.

Reply:
The original code did object allocation, replacement and delete by hand: https://github.com/PowerDNS/pdns/commit/f26bf5471ce15c1e826a8a3250954c2bc26e7370 . That code needed to save the old pointer to be able to delete the old object.

The extra shared pointer ref in the current code indeed causesthe object pointed to by  `t_allowFrom` is only deleted after the `nullptr` assignment in this thread. I don't think this is really needed anymore, but OTOH it does not hurt a lot.

Reply:
I've removed these.